### PR TITLE
feat(test toolkit): add contracts documentation

### PR DESCRIPTION
**Pull request by alachambre**

### CODE REVIEW

File: modules/process-testing/pages/contract.adoc [hunk 1]
```diff
@@ -1,3 +1,112 @@
-= Contracts
+= Contracts in integration tests
+:description: Learn to create contracts to start processes or execute tasks in integration tests using the Bonita test toolkit.
 
-_Coming soon_
\ No newline at end of file
+{description} +
+There are two ways to create contracts: using the Java builder or from a JSon file in the classpath. The Java builder is handy for small contracts but can become tedious for large and complex contracts.
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]: Emphasis:
- one can be used for dynamic values, not the other

File: modules/process-testing/pages/contract.adoc [hunk 2]
```diff
@@ -1,3 +1,112 @@
...
+[NOTE]
+====
+Integrations tests are built upon the Bonita Test Toolkit, based on the open source https://github.com/bonitasoft/bonita-java-client[Bonita Java Client]. +
+The Bonita Test Toolkit is only available for Enterprise, Performance, Efficiency and Teamwork. editions. 
+====
+
+== Create contracts using the Java builder
+
+The Java builder allows to create contracts by specifying contract inputs one by one. +
+It is a pretty handy way to create small contracts, but for large contracts it is recommended to store theme in dedicated JSon files and to retrieve them xref:_retrieve_contracts_from_files[from the classpath].
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]: large and static

File: modules/process-testing/pages/contract.adoc [hunk 3]
```diff
@@ -1,3 +1,112 @@
...
+    .textInput("textContractInput", "value")
+    .multipleTextInput("multipleTextCntractInput", List.of("value1", "value2"))
+    .integerInput("intergerContractInput", 1)
+    .decimalInput("decimalContractInput", 1.0)
+    .booleanInput("booleanContractInput", true)
+    .dateInput("dateContractInput", new Date())
+    .localDateInput("localDateContractInput", LocalDate.now())
+    .localDateTimeInput("localDateTimeContractInput", LocalDateTime.now())
+    .offsetDateTimeInput("offsetDateTimeContractInput", OffsetDateTime.now())
+    .fileInput("fileContractInput", "classpathURL") <1>
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]:
```suggestion
    .fileInput("fileContractInput", "classpathResourceURL") <1>
```

File: modules/process-testing/pages/contract.adoc [hunk 4]
```diff
@@ -1,3 +1,112 @@
...
+    .multipleDecimalInput("multipleDecimalContractInput", List.of(1.0, 2.0))
+    .multipleBooleanInput("multipleBooleanContractInput", List.of(true, false))
+    .multipleDateInput("multipleDateContractInput", List.of(new Date(), new Date()))
+    .multipleLocalDateInput("multipleLocalDateContractInput", List.of(LocalDate.now(), LocalDate.now()))
+    .multipleLocalDateTimeInput("multipleLocalDateTimeContractInput", List.of(LocalDateTime.now(), LocalDateTime.now()))
+    .multipleOffsetDateTimeInput("multipleOffsetDateTimeContractInput", List.of(OffsetDateTime.now(), OffsetDateTime.now()))
+    .multipleFileInput("multipleFileContractInput", List.of("classpathURL1", "classpathURL2"))
+    .build();
+----
+<1> For file contract inputs, the file to use must be added to the classpath, and the corresponding path has to be used in the contratc builder. For example, put your files in `src/test/resources/documents` and use the URL `/documents/<FILE NAME>`.
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]:
```suggestion
<1> For file contract inputs, the file to use must be added to the classpath, and the corresponding path has to be used in the contratc builder. For example, put your files in `src/test/resources/documents` and retrieve the resource URL like this: `MyTest.class.getResource("/documents/<FILE NAME>")`.
```

File: modules/process-testing/pages/contract.adoc [hunk 5]
```diff
@@ -1,3 +1,112 @@
...
+
+=== Complex contract inputs
+
+A complex contract input has for value a list of contract inputs (simple or complex), It can be seen as a tree. +
+It is built like a simple contract input, but with a dedicated layer for the value :
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]:
```suggestion
It is built like a simple contract input, but with a dedicated builder for the value : 
```
?

File: modules/process-testing/pages/contract.adoc [hunk 6]
```diff
@@ -1,3 +1,112 @@
...
+[source, Java]
+----
+ContractBuilder.newContract()
+    .complexInput("complexContractInput", ComplexInputBuilder.complexInput()
+        .localDateInput("localDateContractInput", LocalDate.now())
+        .textInput("textContractInput", "value"))
+    .build();
+----
+
+To build a contract with several layers of complex contract inputs, chain them in the builder:
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]:
```suggestion
To build a contract with several levels of complex contract inputs, chain them in the builder: 
```
?

File: modules/process-testing/pages/contract.adoc [hunk 7]
```diff
@@ -1,3 +1,112 @@
...
+----
+ContractBuilder.newContract()
+    .complexInput("parentComplexContractInput", ComplexInputBuilder.complexInput() <1>
+        .localDateInput("localDateContractInput", LocalDate.now())
+        .textInput("textContractInput", "value")
+        .complexInput("childComplexContractInput", ComplexInputBuilder.complexInput() <2>
+            .booleanInput("booleanContractInput", true)))
+    .build();
+----
+<1> The first layer of contract input, the parent complex contract input which contains all the other contract inputs.
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]:
```suggestion
<1> The first level of contract input, the parent complex contract input which contains all the other contract inputs.
```

File: modules/process-testing/pages/contract.adoc [hunk 8]
```diff
@@ -1,3 +1,112 @@
...
+<2> The second layer of contract input, a child complex contract input which contains other contract inputs.
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]:
```suggestion
<2> The second level of contract input, a child complex contract input which contains other contract inputs.
```

File: modules/process-testing/pages/contract.adoc [hunk 9]
```diff
@@ -1,3 +1,112 @@
...
+
+Has you can see, if a contract contains several layers of complex contract inputs with a lot of fields, using the Java builder to create the contract can become tedious. For such situations, it is recommended to extract the comtract in a dedicated JSon file and to retrieve them xref:_retrieve_contracts_from_files[from the classpath].
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]:
```suggestion
If a contract contains several levels of complex contract inputs with a lot of fields, using the Java builder to create the contract can become tedious. When a contract only contains static values, you may extract the contract in an external JSon file and retrieve them xref:_retrieve_contracts_from_files[from the classpath] to keep the test class concise.
```

File: modules/process-testing/pages/contract.adoc [hunk 10]
```diff
@@ -1,3 +1,112 @@
...
+Contracts can be stored in JSon files, and then retrieve at runtime from the classpath resources. This allows to extract the contract definition and the values in a dedicated file, which is easier to read and to maintain, especially for large and complex contracts. 
+
+A contract JSon file works as a key-value model, the key is the name of the contract input and the value the value of the contract input:
+
+[source, JSon]
+----
+{
+    "textContractInput" : "value",
+    "booleanContractInput" : true,
+    "dateContractInput" : "2021-10-10T00:00:00Z",
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]: Explicit the use of the iso-8601 standard format in a call out

**Review by rbioteau — CHANGES_REQUESTED** [review]

**Comment by rbioteau** [comment]
Can we add an additional section on how to test constraint using `StartProcessException` and `ExecuteTaskException` message ?

```java
        var invalidContract = newContract()
                .complexInput("ticketInput", complexInput().textInput("description", "Some description"))
                .multipleFileInput("attachmentsInput", List.of())
                .build();

        var process = toolkit.getProcessDefinition(NEW_TICKET_PROCESS_NAME);
        assertThatThrownBy(() -> process.startProcessFor(helen, invalidContract))
                .isInstanceOf(StartProcessException.class)
                .hasMessageContaining("Expected input [title] is missing");
```

**Review by rbioteau — COMMENTED** [review]
Can we add an additional section on how to test constraint using `StartProcessException` and `ExecuteTaskException` message ?

```java
        var invalidContract = newContract()
                .complexInput("ticketInput", complexInput().textInput("description", "Some description"))
                .multipleFileInput("attachmentsInput", List.of())
                .build();

        var process = toolkit.getProcessDefinition(NEW_TICKET_PROCESS_NAME);
        assertThatThrownBy(() -> process.startProcessFor(helen, invalidContract))
                .isInstanceOf(StartProcessException.class)
                .hasMessageContaining("Expected input [title] is missing");
```

**Comment by alachambre** [comment]
OKay chef

File: modules/process-testing/pages/contract.adoc [hunk 11]
```diff
@@ -1,3 +1,146 @@
...
+
+        var invalidContract = ContractBuilder.newContract()
+            .complexInput("personContractInput", ComplexInputBuilder.complexInput()
+                .textInput("name", "Adrien"))
+                .localDateInput("birthdate", LocalDate.now().plusDays(1))
+            .build();
+
+        assertThatThrownBy(() -> processDef.startProcessFor(walter, invalidContract))
+                        .isInstanceOf(StartProcessException.class)
+                        .hasMessageContaining("The input [birthdate] cannot be in the futur!");
```
Comment (rbioteau) [inline modules/process-testing/pages/contract.adoc]: add a call-out explaining that the message is defined in the technical message field of the constraint

**Review by rbioteau — CHANGES_REQUESTED** [review]

**Pull request merged**